src/torrent.cpp

```diff
@@ -5078,33 +5096,33 @@ bool is_downloading_state(int const st)
 		// setting file priority on a torrent that doesn't have metadata yet is
 		// similar to having passed in file priorities through add_torrent_params.
 		// we store the priorities in m_file_priority until we get the metadata
-		file_storage const& fs = m_torrent_file->files();
-		if (index < file_index_t(0) || (valid_metadata() && index >= fs.end_file()))
+		if (index < file_index_t(0)
+			|| (valid_metadata() && index >= m_torrent_file->files().end_file()))
 		{
 			return;
 		}
 
 		prio = aux::clamp(prio, dont_download, top_priority);
-		if (m_file_priority.end_index() <= index)
+		auto new_priority = m_file_priority;
+		if (new_priority.end_index() <= index)
 		{
 			// any unallocated slot is assumed to have the default priority
-			if (prio == default_priority) return;
-			m_file_priority.resize(static_cast<int>(index) + 1, default_priority);
+			new_priority.resize(static_cast<int>(index) + 1, default_priority);
 		}
 
-		if (m_file_priority[index] == prio) return;
-		m_file_priority[index] = prio;
-
-		if (!valid_metadata()) return;
+		new_priority[index] = prio;
 
 		// storage may be nullptr during shutdown
 		if (m_storage)
 		{
 			ADD_OUTSTANDING_ASYNC("file_priority");
 			m_ses.disk_thread().async_set_file_priority(m_storage
-				, m_file_priority, std::bind(&torrent::on_file_priority, shared_from_this(), _1, _2));
+				, std::move(new_priority), std::bind(&torrent::on_file_priority, shared_from_this(), _1, _2));
+		}
+		else
+		{
+			m_file_priority = new_priority;
```

Comment:
Using `swap` is more effective then copy assignment. Using move assignment is also fine.